docs/designs/supporting_pytorch.md

```diff
@@ -0,0 +1,297 @@
+Epoch:  1 | Step:  0 | batch x:  tensor([1., 6., 7., 5., 2., 9., 8.]) | batch y:  tensor([10.,  5.,  4.,  6.,  9.,  2.,  3.])
+Epoch:  1 | Step:  1 | batch x:  tensor([10.,  3.,  4.]) | batch y:  tensor([1., 8., 7.])
+Epoch:  2 | Step:  0 | batch x:  tensor([ 4.,  1.,  9.,  5.,  7.,  8., 10.]) | batch y:  tensor([ 7., 10.,  2.,  6.,  4.,  3.,  1.])
+Epoch:  2 | Step:  1 | batch x:  tensor([3., 2., 6.]) | batch y:  tensor([8., 9., 5.])
+```
+
+```python
+import torch
+import torch.utils.data as Data
+
+BATCH_SIZE = 5
+x = torch.linspace(1, 10, 10)       # this is x data (torch tensor)
+y = torch.linspace(10, 1, 10)       # this is y data (torch tensor)
+
+torch_dataset = Data.TensorDataset(x, y)
+loader = Data.DataLoader(
+    dataset=torch_dataset,      # torch TensorDataset format
+    batch_size=BATCH_SIZE,      # mini batch size
+    shuffle=True,               # random shuffle for training
+    num_workers=2,              # subprocesses for loading data
+)
+def show():
+    for epoch in range(3):   # train entire dataset 3 times
+        for step, (batch_x, batch_y) in enumerate(loader):  # for each training step
+            # train your data...
+            print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            batch_x, '| batch y: ', batch_y)
+            # print('Epoch: ', epoch, '| Step: ', step, '| batch x: ',
+            #       batch_x.numpy(), '| batch y: ', batch_y.numpy())
+if __name__ == '__main__':
+    show()
+```
+### MNIST Example
+
+This dataset is in numpy array format, and has been stored using pickle, a python-specific format for serializing data. Each image is 28 x 28, and is being stored as a flattened row of length 784 (=28x28). Let’s take a look at one; we need to reshape it to 2d first.
+```python
+plt.imshow(train_data.train_data[0].numpy(), cmap='gray')
+plt.title('%i' % train_data.train_labels[0])
+plt.show()
+```
+
+PyTorch uses `torch.tensor`, rather than `numpy` arrays.
+```python
+x_train, y_train, x_valid, y_valid = map(
+    torch.tensor, (x_train, y_train, x_valid, y_valid)
+)
+```
+Or we can convert our data like this.
+```python
+train_data = torchvision.datasets.MNIST(
+    root='./mnist/',
+    train=True,                                     # this is training data
+    transform=torchvision.transforms.ToTensor(),    # Converts a PIL.Image or numpy.ndarray to
+                                                    # torch.FloatTensor of shape (C x H x W) and normalize in the range [0.0, 1.0]
+    download=DOWNLOAD_MNIST,
+)
+```
+
+
+
+## Training Loop 
+ElasticDL holds master-worker architecture. Workers get model parameters from parameter server (PS), compute gradients using different training data, and send computed gradients to PS. PS iteratively updates these model parameters using gradients sent by workers.
+In our training loop, the gradient information of the neural network needs to be calculated and transmitted between the master and the worker.
```

Comment:
Why should we transmit gradients between the master and workers?

Reply:
I made a mistake. It is ps rather than master.